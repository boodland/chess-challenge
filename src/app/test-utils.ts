const PIECES = 'KQRBNP';

export const PieceShape = {
    'K': '♔',
    'Q': '♕',
    'R': '♖',
    'B': '♗',
    'N': '♘',
    'P': '♙',
    'k': '♚',
    'q': '♛',
    'r': '♜',
    'b': '♝',
    'n': '♞',
    'p': '♟',
    '': '',
};

export const EMPTY_PIECE = PieceShape[''];

const getRandomPiece = () => PIECES.charAt(Math.floor(Math.random() * PIECES.length));
export const getRandomWhitePiece = getRandomPiece;
export const getRandomBlackPiece = () => getRandomPiece().toLowerCase();

export const EMPTY_HANDLER = () => null;

