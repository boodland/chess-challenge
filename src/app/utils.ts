export const COLUMNS = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
export const ROWS = [0, 1, 2, 3, 4, 5, 6, 7];
export const FEN_ROWS_SEPARATOR = "/";
export const INITIAL_FEN = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR';
export const LOCAL_STORAGE_KEY = "fen";

export enum CellTypeEnum {
    light = "light",
    dark = "dark",
};

enum PieceShapeEnum {
    'K' = '♔',
    'Q' = '♕',
    'R' = '♖',
    'B' = '♗',
    'N' = '♘',
    'P' = '♙',
    'k' = '♚',
    'q' = '♛',
    'r' = '♜',
    'b' = '♝',
    'n' = '♞',
    'p' = '♟',
    '' = '',
}

export function getColor(piece: string) {
    return (piece === piece.toUpperCase()) ? "white" : "black";
}

export function areSameColor(piece1: string, piece2: string) {
    return !isEmpty(piece1) && !isEmpty(piece2) && getColor(piece1) === getColor(piece2);
}

export function getShape(piece: string) {
    return PieceShapeEnum[piece as keyof typeof PieceShapeEnum];
}

export function isEmpty(piece: string) {
    return piece === "";
}

export function getFromLocalStorage(): string {
    const localFen = localStorage.getItem(LOCAL_STORAGE_KEY);
    return localFen || INITIAL_FEN;
}

export function saveToLocalStorage(fen: string) {
    localStorage.setItem(LOCAL_STORAGE_KEY, fen);
}

export function isFenValid(fen: string) {
    let isValid = isFenWellConstructed(fen);
    isValid = isValid && areNumberOfItemsInSectionsCorrect(fen);
    isValid = isValid && areNumberOfPiecesCorrect(fen);
    return isValid;
}

function isFenWellConstructed(fen: string) {
    const isValid = /^(?!.*\d{2,}.*)([1-8PNBRQK]+\/){7}[1-8PNBRQK]+$/gim.test(fen);
    return isValid;
}

function areNumberOfItemsInSectionsCorrect(fen: string) {
    const fenSections = fen.split(FEN_ROWS_SEPARATOR);
    let isValid =  true;
    fenSections.forEach((section) => {
        isValid = isValid && isNumberOfItemsInSectionCorrect(section);
    });
    return isValid;
}

function isNumberOfItemsInSectionCorrect(section: string) {
    const sectionChars = [...section];
    const total =  sectionChars.reduce((total, char) => {
        let increment = parseInt(char);
        if(isNaN(increment)) increment = 1;
        return total + increment;
    }, 0);
    return total === 8;
}

function areNumberOfPiecesCorrect(fen: string) {
    const fenChars = [...fen];
    let isValid = isNumberOfPieceCorrect(fenChars, 'p', 8);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'P', 8);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'r', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'R', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'n', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'N', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'b', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'B', 2);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'k', 1);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'K', 1);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'Q', 1);
    isValid = isValid && isNumberOfPieceCorrect(fenChars, 'q', 1);
    return isValid;
}

function isNumberOfPieceCorrect(fen: string[], piece: string, maximum: number) {
    const total = fen.filter((char) => char === piece).length;
    return total <= maximum;
}