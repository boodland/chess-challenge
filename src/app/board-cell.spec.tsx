import { fireEvent, getByRole, render } from '@testing-library/react';

import BoardCell from './board-cell';
import { CellTypeEnum } from './utils';
import { EMPTY_HANDLER, EMPTY_PIECE, PieceShape } from './test-utils';

describe('BoardCell', () => {
  it('should render successfully', () => {
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.dark}
        onClick={EMPTY_HANDLER}
      />
    );
    expect(container).toBeTruthy();
  });

  it('should display light cell when type is light', () => {
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.light}
        onClick={EMPTY_HANDLER}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should display dark cell when type is dark', () => {
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.dark}
        onClick={EMPTY_HANDLER}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should display blocked cell when is blocked', () => {
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.dark}
        blocked={true}
        onClick={EMPTY_HANDLER}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should display selected cell when is selected', () => {
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.dark}
        selected={true}
        onClick={EMPTY_HANDLER}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should display the correct piece text for each piece', () => {
    Object.entries(PieceShape).forEach(([piece, shape]) => {
      const { container } = render(
        <BoardCell
          piece={piece}
          type={CellTypeEnum.dark}
          onClick={EMPTY_HANDLER}
        />
      );
      const button = getByRole(container, 'button');
      expect(button.textContent).toEqual(shape);
    });
  });

  it('should call onCellClick when clicking the cell', () => {
    const onCellClickSpy = vi.fn();
    const { container } = render(
      <BoardCell
        piece={EMPTY_PIECE}
        type={CellTypeEnum.light}
        onClick={onCellClickSpy}
      />
    );
    const button = getByRole(container, 'button');
    fireEvent.click(button);
    expect(onCellClickSpy).toHaveBeenCalledOnce();
  });
});
