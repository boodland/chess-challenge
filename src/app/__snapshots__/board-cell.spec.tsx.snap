// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoardCell > should display blocked cell when is blocked 1`] = `
<div>
  <button
    class="_cell_436b5f _dark_436b5f _blocked_436b5f "
  />
</div>
`;

exports[`BoardCell > should display dark cell when type is dark 1`] = `
<div>
  <button
    class="_cell_436b5f _dark_436b5f  "
  />
</div>
`;

exports[`BoardCell > should display light cell when type is light 1`] = `
<div>
  <button
    class="_cell_436b5f _light_436b5f  "
  />
</div>
`;

exports[`BoardCell > should display selected cell when is selected 1`] = `
<div>
  <button
    class="_cell_436b5f _dark_436b5f  _selected_436b5f"
  />
</div>
`;
